virtool/samples/data.py

```diff
@@ -474,3 +475,23 @@ async def deduplicate_sample_names(self):
                         {"$set": {"name": await name_generator.get(session)}},
                         session=session,
                     )
+
+    async def populate_workflows_field(self):
+        cursor = self._db.samples.find({"workflows": {"$exists": False}})
+        samples = await cursor.to_list(length=None)
+
+        sample_ids = [samples["_id"] for sample in samples]
+
+        if sample_ids:
+            for sample_id in sample_ids:
+                library_type = await get_one_field(
+                    self._db.samples, "library_type", sample_id
+                )
+                analyses = await self._db.analyses.find(
+                    {"sample.id": sample_id}, ["ready", "workflow"]
+                ).to_list(None)
+
+                workflow_states = derive_workflow_state(analyses, library_type)
+                await self._db.samples.find_one_and_update(
+                    {"_id": sample_id}, {"$set": {"workflows": workflow_states}}
+                )
```

Comment:
1. Iterate over the cursor. Using `to_list` loads all of the matching documents into memory, which could 
2. You are only using `_id` at line 483. Fetching the entire sample documents for one string is wasteful.
3. No need for `if`. The iteration will not happen on a zero-length list.
4. Use `get_one_field` is redundant. You already have this field in hand at line 480.
5. At line 495. Just use `update_one` if you are not using the returned updated document. This avoid unnecessary data transimission.

```suggestion
        async for sample in self._db.samples.find({"workflows": {"$exists": False}}, ["_id", "library_type"]):
            sample_id = sample["_id"]
            
            analyses = await self._db.analyses.find(
                {"sample.id": sample_id}, ["ready", "workflow"]
            ).to_list(None)

            workflow_states = derive_workflow_state(analyses, sample["library_type"])
            
            await self._db.samples.find_one_and_update(
                {"_id": sample_id}, {"$set": {"workflows": workflow_states}}
            )
```